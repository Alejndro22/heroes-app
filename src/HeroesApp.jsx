import { Outlet } from 'react-router';
import { Navbar } from './ui';

export const HeroesApp = () => {
  return (
    <>
      <Navbar />
      <Outlet />
    </>
  );
};
