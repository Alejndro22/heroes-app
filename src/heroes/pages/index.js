export * from './DcPage';
export * from './ErrorPage';
export * from './HeroPage';
export * from './MarvelPage';
export * from './SearchPage';
