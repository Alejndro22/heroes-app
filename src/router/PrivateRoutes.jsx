import { useContext, useEffect } from 'react';
import { Navigate, useLocation } from 'react-router-dom';
import { AuthContext } from '../auth';

// eslint-disable-next-line react/prop-types
export const PrivateRoutes = ({ children }) => {
  const { logged } = useContext(AuthContext);

  // useLocation hook returns the current location elements
  const { pathname, search } = useLocation();

  useEffect(() => {
    const lastPath = pathname + search;
    localStorage.setItem('lastPath', lastPath);
  }, [pathname, search]);

  if (pathname + search === '/') return <Navigate to={'/marvel'} />;

  return logged ? children : <Navigate to={'/login'} />;
};
