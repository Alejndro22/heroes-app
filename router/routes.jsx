import { HeroesApp } from '../src/HeroesApp';
import { LoginPage } from '../src/auth/pages/LoginPage';
import { DcPage } from '../src/heroes/pages/DcPage';
import { MarvelPage } from '../src/heroes/pages/MarvelPage';
import ErrorPage from '../src/heroes/pages/ErrorPage';

export const routes = [
  {
    path: '/',
    element: <HeroesApp />,
    errorElement: <ErrorPage />,
    children: [
      {
        path: 'marvel',
        element: <MarvelPage />,
      },
      {
        path: 'dc',
        element: <DcPage />,
      },
      {
        path: 'login',
        element: <LoginPage />,
      },
    ],
  },
];
