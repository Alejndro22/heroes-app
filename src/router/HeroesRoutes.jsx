// import { Navigate } from 'react-router-dom';
import { DcPage, HeroPage, MarvelPage, SearchPage } from '../heroes';

export const heroesChildrenRoutes = [
  {
    path: 'marvel',
    element: <MarvelPage />,
  },
  {
    path: 'dc',
    element: <DcPage />,
  },
  {
    path: 'search',
    element: <SearchPage />,
  },
  {
    path: 'hero/:id',
    element: <HeroPage />,
  },
  // {
  //   path: '/*',
  //   element: <Navigate to={'/marvel'} replace={true} />,
  // },
];
