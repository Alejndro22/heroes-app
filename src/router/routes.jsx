import { HeroesApp } from '../HeroesApp';
import { LoginPage } from '../auth';
import { DcPage, MarvelPage } from '../heroes';
import ErrorPage from '../heroes/pages/ErrorPage';

export const routes = [
  {
    path: '/',
    element: <HeroesApp />,
    errorElement: <ErrorPage />,
    children: [
      {
        path: 'marvel',
        element: <MarvelPage />,
      },
      {
        path: 'dc',
        element: <DcPage />,
      },
      {
        path: 'login',
        element: <LoginPage />,
      },
    ],
  },
];
