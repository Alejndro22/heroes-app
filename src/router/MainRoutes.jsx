import { HeroesRoutes } from '../heroes';
import { LoginPage } from '../auth';
import { heroesChildrenRoutes } from './HeroesRoutes';
import ErrorPage from '../heroes/pages/ErrorPage';
import { PrivateRoutes } from './PrivateRoutes';
import { PublicRoutes } from './PublicRoutes';

export const routes = [
  {
    path: '/login',
    element: (
      <PublicRoutes>
        <LoginPage />
      </PublicRoutes>
    ),
  },

  {
    path: '/',
    element: (
      <PrivateRoutes>
        <HeroesRoutes />
      </PrivateRoutes>
    ),
    errorElement: <ErrorPage />,
    children: heroesChildrenRoutes,
  },

  // For reference, this works when there's no need to check private/public routes
  // { path: 'login', element: <LoginPage /> },
];
