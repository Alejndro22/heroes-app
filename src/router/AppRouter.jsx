import { createBrowserRouter, RouterProvider } from 'react-router-dom';
import { routes } from './MainRoutes';

const router = createBrowserRouter(routes);

export const AppRouter = () => {
  return <RouterProvider router={router} />;
};
