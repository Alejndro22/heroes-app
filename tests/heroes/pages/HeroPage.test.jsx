import {
  MemoryRouter,
  RouterProvider,
  Route,
  Routes,
  createMemoryRouter,
} from 'react-router-dom';
import { AuthContext } from '../../../src/auth';
import { routes } from '../../../src/router/MainRoutes';
import { HeroPage } from '../../../src/heroes';
import { fireEvent, render, screen } from '@testing-library/react';

// Mock useNavigate hook
const mockedUseNavigate = jest.fn();
// Here i tell to keep everything as default, and only useNavigate() will be mocked
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUseNavigate,
}));

describe('tests on <HeroPage />', () => {
  beforeEach(() => jest.clearAllMocks());

  const contextValue = {
    logged: true,
    user: { id: 'abc', name: 'testing' },
  };

  test('should match snapshot', () => {
    // Here I used MemoryRouter because i didn't want to evaluate the Outlet logic,
    // just the page by itself
    const { container } = render(
      <AuthContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={['/hero/dc-arrow']}>
          <Routes>
            <Route path='/hero/:id' element={<HeroPage />} />
          </Routes>
        </MemoryRouter>
      </AuthContext.Provider>
    );

    expect(container).toMatchSnapshot();
  });

  test('should render page by Hero if found', () => {
    const router = createMemoryRouter(routes, {
      initialEntries: ['/hero/dc-arrow'],
    });

    render(
      <AuthContext.Provider value={contextValue}>
        <RouterProvider router={router}>
          <HeroPage />
        </RouterProvider>
      </AuthContext.Provider>
    );

    const heroTitle = screen.getByRole('heading', { level: 3 });
    const img = screen.getByRole('img');

    expect(heroTitle.innerHTML).toEqual('Green Arrow');
    expect(img.src).toContain('/assets/heroes/dc-arrow.jpg');
  });

  test('should Navigate to /marvel if hero not found', () => {
    const router = createMemoryRouter(routes, {
      initialEntries: ['/hero/hero-not-found'],
    });

    render(
      <AuthContext.Provider value={contextValue}>
        <RouterProvider router={router}>
          <HeroPage />
        </RouterProvider>
      </AuthContext.Provider>
    );

    expect(screen.getByText('Marvel Comics')).toBeTruthy();
  });

  test('should Navigate to last path visited when button is clicked', () => {
    const router = createMemoryRouter(routes, {
      initialEntries: ['/hero/dc-arrow'],
    });

    render(
      <AuthContext.Provider value={contextValue}>
        <RouterProvider router={router}>
          <HeroPage />
        </RouterProvider>
      </AuthContext.Provider>
    );

    const backBtn = screen.getByRole('button', { name: 'Back' });
    fireEvent.click(backBtn);

    expect(mockedUseNavigate).toHaveBeenCalledWith(-1);
  });
});
