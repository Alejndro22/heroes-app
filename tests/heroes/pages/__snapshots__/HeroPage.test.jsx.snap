// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on <HeroPage /> should match snapshot 1`] = `
<div>
  <div
    class="mt-8 grid grid-cols-3 gap-8 animate__animated animate__fadeInLeft"
  >
    <div
      class="col-span-1"
    >
      <img
        alt="Green Arrow"
        class="border border-gray-300 rounded-md p-1 h-64 sm:h-auto"
        src="./../assets/heroes/dc-arrow.jpg"
      />
    </div>
    <div
      class="col-span-2"
    >
      <h3
        class="mb-4 text-2xl font-semibold tracking-tight text-gray-900 md:text-3xl lg:text-4xl"
      >
        Green Arrow
      </h3>
      <ul
        class="list-none text-sm space-y-2"
      >
        <li>
          <b>
            Alter ego: 
          </b>
          Oliver Queen
        </li>
        <hr />
        <li>
          <b>
            Publisher: 
          </b>
          DC Comics
        </li>
        <hr />
        <li>
          <b>
            First appearance: 
          </b>
          More Fun Comics #73
        </li>
        <hr />
      </ul>
      <h5
        class="mt-3 text-lg font-semibold text-gray-900 md:text-xl lg:text-2xl"
      >
        Characters
      </h5>
      <p
        class="text-sm my-2"
      >
        Oliver Queen
      </p>
      <button
        class="mt-2 bg-transparent hover:bg-emerald text-emerald font-semibold hover:text-white py-1 px-4 border border-emerald hover:border-transparent rounded"
      >
        Back
      </button>
    </div>
  </div>
</div>
`;
